### PR TITLE
Update check and inject output

### LINKED ISSUE #1471: Revisit linkerd check output format
Per #1081, we should streamline the linkerd check output to be less verbose when checks are passing, and more verbose when checks are failing. We should also group the checks as Pre-Install and Post-Install. Relates to #1470.

### PR DESCRIPTION
**Pull request by siggy**
The outputs of the `check` and `inject` commands did not vary much
between successful and failed executions, and were a bit verbose and
challenging to parse.

Reorganize output of `check` and `inject` commands, to provide more
output when errors occur, and less output when successful.

Specific changes:

`linkerd check`
- visually group checks by category
- introduce `hintURL`'s, to provide doc links when checks fail
- add spinners when retrying, remove addtional retry lines
- colored emojis to indicate success/warning/failure

`linkerd inject`
- modify default output to mirror `kubectl apply`
- only output non-successful inject reports
- support `--verbose` flag to output all inject reports

Fixes #1471, #1653, #1656, #1739

Signed-off-by: Andrew Seigner <siggy@buoyant.io>

# Example outputs

Note the unicode symbols are rendered with color in the terminal:

screen shot 2019-01-15 at 3 35 08 pm

```
✔
✘
⚠
```

## `linkerd check` with retries

check-spinner4

## `linkerd check --pre`

```
$ bin/linkerd check --pre --expected-version=$(bin/root-tag)
kubernetes-api
--------------
✔ can initialize the client
✔ can query the Kubernetes API

kubernetes-version
------------------
✔ is running the minimum Kubernetes API version

pre-kubernetes-cluster-setup
----------------------------
✘ control plane namespace does not already exist
    The "linkerd" namespace already exists
✔ can create Namespaces
✔ can create ClusterRoles
✔ can create ClusterRoleBindings
✔ can create CustomResourceDefinitions

pre-kubernetes-setup
--------------------
✔ can create ServiceAccounts
✔ can create Services
✔ can create Deployments
✔ can create ConfigMaps

linkerd-version
---------------
✔ can determine the latest version
✔ cli is up-to-date

Status check results are ✘
```

## `linkerd check`
```
$ bin/linkerd check --expected-version=$(bin/root-tag)
kubernetes-api
--------------
✔ can initialize the client
✔ can query the Kubernetes API

kubernetes-version
------------------
✔ is running the minimum Kubernetes API version

linkerd-existence
-----------------
✔ control plane namespace exists
✔ controller pod is running
✔ can initialize the client
✔ can query the control plane API

linkerd-api
-----------
✔ control plane pods are ready
✔ can query the control plane API
✔ [kubernetes] control plane can talk to Kubernetes
✔ [prometheus] control plane can talk to Prometheus

linkerd-service-profile
-----------------------
✔ no invalid service profiles

linkerd-version
---------------
✔ can determine the latest version
✔ cli is up-to-date

control-plane-version
---------------------
⚠ control plane is up-to-date
    is running version 19.1.1 but the latest dev version is 75971c4b-siggy

Status check results are ✔
```

## `linkerd check --pre` with errors

```
$ bin/linkerd check --pre --expected-version=$(bin/root-tag)
kubernetes-api
--------------
✔ can initialize the client
✔ can query the Kubernetes API

kubernetes-version
------------------
✔ is running the minimum Kubernetes API version

pre-kubernetes-cluster-setup
----------------------------
✘ control plane namespace does not already exist
    The "linkerd" namespace already exists
✔ can create Namespaces
✘ can create ClusterRoles
    Missing permissions to create ClusterRoles
    See https://linkerd.io/2/getting-started/#step-0-setup for hints
✔ can create ClusterRoleBindings
✔ can create CustomResourceDefinitions

pre-kubernetes-setup
--------------------
✔ can create ServiceAccounts
✔ can create Services
✔ can create Deployments
✔ can create ConfigMaps

linkerd-version
---------------
✔ can determine the latest version
✔ cli is up-to-date

Status check results are ✘
```

## `linkerd check --proxy`

```
$ bin/linkerd check --expected-version=$(bin/root-tag) --proxy
kubernetes-api
--------------
✔ can initialize the client
✔ can query the Kubernetes API

kubernetes-version
------------------
✔ is running the minimum Kubernetes API version

linkerd-existence
-----------------
✔ control plane namespace exists
✔ controller pod is running
✔ can initialize the client
✔ can query the control plane API

linkerd-api
-----------
✔ control plane pods are ready
✔ can query the control plane API
✔ [kubernetes] control plane can talk to Kubernetes
✔ [prometheus] control plane can talk to Prometheus

linkerd-service-profile
-----------------------
✔ no invalid service profiles

linkerd-version
---------------
✔ can determine the latest version
✔ cli is up-to-date

linkerd-data-plane
------------------
✔ data plane namespace exists
✔ data plane proxies are ready
✔ data plane proxy metrics are present in Prometheus
⚠ data plane is up-to-date
    linkerd/linkerd-prometheus-568f746f6b-pvs6w is running version edge-19.1.1 but the latest version is dev-75971c4b-siggy

Status check results are ✔
```

## `linkerd inject`

```
$ curl -s https://run.linkerd.io/emojivoto.yml | bin/linkerd inject - | kubectl apply -f -

namespace "emojivoto" skipped
deployment "emoji" injected
service "emoji-svc" skipped
deployment "voting" injected
service "voting-svc" skipped
deployment "web" injected
service "web-svc" skipped
deployment "vote-bot" injected

namespace "emojivoto" created
deployment.apps "emoji" created
service "emoji-svc" created
deployment.apps "voting" created
service "voting-svc" created
deployment.apps "web" created
service "web-svc" created
deployment.apps "vote-bot" created
```

## `linkerd inject --verbose`

```
$ curl -s https://run.linkerd.io/emojivoto.yml | bin/linkerd inject --verbose - | kubectl apply -f -

✔ pods do not use host networking
✔ pods do not have a proxy or initContainer already injected
✔ at least one resource injected
✔ pod specs do not include UDP ports

namespace "emojivoto" skipped
deployment "emoji" injected
service "emoji-svc" skipped
deployment "voting" injected
service "voting-svc" skipped
deployment "web" injected
service "web-svc" skipped
deployment "vote-bot" injected

namespace "emojivoto" created
deployment.apps "emoji" created
service "emoji-svc" created
deployment.apps "voting" created
service "voting-svc" created
deployment.apps "web" created
service "web-svc" created
deployment.apps "vote-bot" created
```

### CODE REVIEW

**Comment by wmorgan** [comment]
I know the Helm folks ran into issue with emoji output in their CLIs -- maybe @michelleN has an opinion?

File: cli/cmd/check.go [hunk 1]
```diff
@@ -150,30 +135,49 @@ func (o *checkOptions) validate() error {
...
+			fmt.Fprintln(w, underline)
+
+			lastCategory = result.Category
 		}
 
+		spin.Stop()
 		if result.Retry {
-			fmt.Fprintf(w, "%s%s%s -- %s%s", checkLabel, filler, retryStatus, result.Err, lineBreak)
+			spin.Suffix = fmt.Sprintf("  %s -- %s", result.Description, result.Err)
+			spin.Color("bgBlack", "bold", "fgRed")
```
Comment (alpeb) [inline cli/cmd/check.go]: I have a light theme in my console (I know, weird, but I have my reasons :) ) so the spinner doesn't look right. Although I looked a bit into the spinner lib and couldn't find a way to cater for both dark and light themes at the same time :-(

Comment (siggy) [reply cli/cmd/check.go]: @alpeb can you try replacing that `spin.Color(...` line with `spin.Start(...`? maybe the default colors will play better in light and dark themes?

**Comment by siggy** [comment]
@grampelberg 
- will try spinner 21
- going to switch to coloring unicode chars (good idea!)
- re: hint URLs, you are correct we'll fill more in later, this PR adds the functionality but only populates a few of the error messages.

File: cli/cmd/check.go [hunk 2]
```diff
@@ -150,30 +135,49 @@ func (o *checkOptions) validate() error {
...
-		checkLabel := fmt.Sprintf("%s: %s", result.Category, result.Description)
+		if lastCategory != result.Category {
+			if lastCategory != "" {
+				fmt.Fprintln(w)
+			}
+
+			underline := ""
+			for i := 0; i < len(result.Category); i++ {
+				underline = underline + "-"
+			}
```
Comment (klingerf) [inline cli/cmd/check.go]: You can use `strings.Repeat` here:

```diff
diff --git a/cli/cmd/check.go b/cli/cmd/check.go
index 8da69362..4d2c8fc9 100644
--- a/cli/cmd/check.go
+++ b/cli/cmd/check.go
@@ -145,13 +145,8 @@ func runChecks(w io.Writer, hc *healthcheck.HealthChecker) bool {
 				fmt.Fprintln(w)
 			}
 
-			underline := ""
-			for i := 0; i < len(result.Category); i++ {
-				underline = underline + "-"
-			}
-
 			fmt.Fprintln(w, result.Category)
-			fmt.Fprintln(w, underline)
+			fmt.Fprintln(w, strings.Repeat("-", len(result.Category)))
 
 			lastCategory = result.Category
 		}
```

File: cli/cmd/inject_util.go [hunk 3]
```diff
@@ -323,12 +328,3 @@ func walk(path string) ([]io.Reader, error) {
 
 	return in, nil
 }
-
-func getFiller(text string) string {
-	filler := ""
-	for i := 0; i < lineWidth-len(text)-len(okStatus)-len("\n"); i++ {
```
Comment (klingerf) [inline cli/cmd/inject_util.go]: 👏 glad this is going away

**Pull request merged**